src/VirtoCommerce.Platform.Web/Controllers/Api/LicensingController.cs

```diff
@@ -17,29 +18,36 @@ namespace VirtoCommerce.Platform.Web.Controllers.Api
     public class LicensingController : Controller
     {
         private readonly PlatformOptions _platformOptions;
+
         public LicensingController(IOptions<PlatformOptions> platformOptions)
         {
             _platformOptions = platformOptions.Value;
         }
 
         [HttpPost]
         [Route("activateByCode")]
-        public async Task<ActionResult<License>> ActivateByCode([FromBody]string activationCode)
+        public async Task<ActionResult<License>> ActivateByCode([FromBody] string activationCode)
         {
-            License license = null;
+            if (!Regex.IsMatch(activationCode, "^([a-zA-Z_0-9-])+$"))
+            {
+                return BadRequest(new { Error = $"{activationCode} is invalid" });
+            }
 
             using (var httpClient = new HttpClient())
             {
                 var activationUrl = new Uri(_platformOptions.LicenseActivationUrl + activationCode);
                 var httpResponse = await httpClient.GetAsync(activationUrl);
-                if (httpResponse.IsSuccessStatusCode)
+
+                if (!httpResponse.IsSuccessStatusCode)
                 {
-                    var rawLicense = await httpResponse.Content.ReadAsStringAsync();
-                    license = License.Parse(rawLicense, Path.GetFullPath(_platformOptions.LicensePublicKeyPath));
+                    return BadRequest(new { Error = $"License server returns {httpResponse.StatusCode} status code" });
```

Comment:
Seems that with the addition of BadRequests, client-side license should be reworked accordingly, so the user could see a proper message, not just empty error. In fact it is breaking change in the public API behavior, you need to think and check how it behaves after that.
Please check how the new code works with:
https://github.com/VirtoCommerce/vc-platform/blob/5d51ae50fddded58be04a454db8daa65e2b94518/src/VirtoCommerce.Platform.Web/wwwroot/js/app/licensing/license-detail.js#L20